src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/AppleCertificatePal.cs

```diff
@@ -109,7 +111,20 @@ internal sealed partial class AppleCertificatePal : ICertificatePal
 
                 using (keychain)
                 {
-                    return ImportPkcs12(rawData, password, exportable, keychain);
+                    var  ret = ImportPkcs12(rawData, password, exportable, keychain);
+                    if (!persist)
+                    {
+                        // If we used temporary keychain we need to prevent deletion.
+                        // on 10.15+ if keychain is unlinked, certain certificate operations may fail.
+                        bool success = false;
+                        keychain.DangerousAddRef(ref success);
+                        if (success)
+                        {
+                            ((AppleCertificatePal)ret)._tempKeychain = keychain;
```

Comment:
I think you can change `ImportPkcs12` and `ImportPkcs12NonExportable` in `AppleCertificatePal.Pkcs12.cs` to return `AppleCertificatePal` so that the cast is not required.